src/Microsoft.AspNetCore.Routing/Patterns/RoutePatternFactory.cs

```diff
@@ -0,0 +1,698 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Routing.Constraints;
+
+namespace Microsoft.AspNetCore.Routing.Patterns
+{
+    public static class RoutePatternFactory
+    {
+        public static RoutePattern Pattern(IEnumerable<RoutePatternPathSegment> segments)
```

Comment:
Hmm, my first thought is I'd lean towards having `static Create(...)` methods on the various types, e.g. `var myNewRoutePattern = RoutePattern.Create(segements)`, instead of a central static help class

Reply:
The code we wrote for this in 2.1 was like this and upon review I thought it was horrible because you end up needing to know like 8 different types. If you're using these then you'd want to `static using` this class and it's pretty nice/simple (see the tests).

Reply:
I would prefer having having the construction of RoutePattern in one place, in this case RoutePatternFactory where I supply 'rawtext and non-inline defaults and constraints' and removing the creation methods on `RoutePattern` class. Some of these methods can be internal I think.

Reply:
If you guys are really sad about this btw we can make it internal. 

@kichalla  if I understand your suggestion are you asking for me to move the `Parse` methods to this class?

Reply:
> @kichalla if I understand your suggestion are you asking for me to move the Parse methods to this class?

Yes @rynowak